src/merkle/partial_mt/mod.rs

```diff
@@ -74,6 +78,62 @@ impl PartialMerkleTree {
         })
     }
 
+    /// Returns a new [PartialMerkleTree] instantiated with leaves map as specified by the provided
+    /// entries.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain an insufficient set of nodes.
+    pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (NodeIndex, RpoDigest)> + ExactSizeIterator,
+    {
+        // check if the number of leaves can be accommodated by the tree's depth; we use a min
+        // depth of 63 because we consider passing in a vector of size 2^64 infeasible.
+        let mut nodes: BTreeMap<NodeIndex, RpoDigest> = entries.into_iter().collect();
```

Comment:
I think a better way might be to build a `BTreeMap<u8, Vec<(u64, RpoDigest)>` where `u8` is node depth, and `u64` is the index value. This way, after a single iteration over the leaves, we'd have everything sorted by depth, and we won't need to do filtering on line 113 or iterating again to get max depth on line 103.

Reply:
I agree that getting `max_depth` and nodes in a certain layer will indeed become easier and faster, but it turns out to be the only advantages of this approach. Conversion of the `BTreeMap<NodeIndex, RpoDigest>` to the `BTreeMap<u8, Vec<(u64, RpoDigest)>` and back turned out to be not that great (but maybe I just didn't find a good way to do that), and calculating internal nodes with this map will be way more complicated and slow. I'm not sure that using this map will be more effective overall.

Reply:
Why would it be more complicated and slow? Let's assume we have this map, then we should be able to do something like:
```Rust
let layers: BTreeMap<u8, Vec<(u64, RpoDigest)>> = ...

for (depth, layer) in layers.into_iter().rev() {
    for (index_value, hash) in layer {
        let index = NodeIndex::new_unchecked(depth, index_value);
        // and then basically the same code as you have now
    }
}
```
Which part causes the slowdown?

Reply:
I was not right, my bad. I managed to fix the majority of complicated and slow code. The only left is:
- Converting `Iterator<Item = (NodeIndex, RpoDigest)>` to the `BTreeMap<u8, Vec<(u64, RpoDigest)>>` and `BTreeMap<u8, Vec<(u64, RpoDigest)>>` to the `BTreeMap<NodeIndex, RpoDigest>`. Maybe it could be done better, but I didn't find a way to do so
- On the each layer I need to get sibling and parent nodes, so I need to `.find()` the desired value in the vector. It should be faster than filtering, but still expensive

It's indeed should be more efficient.

Reply:
> Converting `Iterator<Item = (NodeIndex, RpoDigest)>` to the `BTreeMap<u8, Vec<(u64, RpoDigest)>>` and `BTreeMap<u8, Vec<(u64, RpoDigest)>>` to the `BTreeMap<NodeIndex, RpoDigest>`. Maybe it could be done better, but I didn't find a way to do so

You should be able to do this simpler with just a single iteration over the list of leaves. It would probably be easier to use a simple loop here (rather than iterator) and build both `leaves` and layer_nodes` at the same time. Also, I'd probably use [entry()](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry) method for updating `layer_nodes`.

> On the each layer I need to get sibling and parent nodes, so I need to `.find()` the desired value in the vector. It should be faster than filtering, but still expensive.

I think you should be able to do it without `.find()` for every sibling. Just use the final `nodes` map as the accumulator right away (this will also save an iteration later on). You might need to keep track of orphan nodes as you process each layer - but this can be probably done via a `BTreeSet` and if the nodes are sorted, shouldn't really impact performance.